### PR TITLE
Add ability to use PREWHERE with row-filter (insecure) [DRAFT]

### PR DESCRIPTION
**Pull request by azat**
**Do not look, do not merge**

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to use PREWHERE with row-filter (insecure, done under option `allow_insecure_prewhere` just as a POC)

Detailed description / Documentation draft:
Implement allow_insecure_prewhere server setting [DRAFT]

Right now row-level security filters disables PREWHERE, due to security
issues, since right now both expression in the and() function will be
executed (due to short circuit evaluation is not supported), i.e.:

    and(row_level_filter, query_expr)

And if query_expr will have throwIf(column from row-level filter) it can
know the column values boundaries (at least), also note that throwIf()
is not the only possible security breach.

But sometimes queries w/o PREWHERE works too damn slow, so this patch
introduces allow_insecure_prewhere, that will allow using row-level
filter with PREWHERE. Basically with this option row-level filter will
be moved into PREWHERE (not WHERE, since this will introduce more
security breaches).

Also note that this is just a DRAFT and should not considered seriosly,
but once short circuit evaluation will be supported this can be
implemented in the right way (although there can be some caveats due to
order will became important).

<details>

HEAD:
- f5c8f92f5d7e81d0380ec1913208da9c08d9838f

</details>

### CODE REVIEW

**Comment by filimonov** [comment]
Not working :\ 

https://github.com/azat/ClickHouse/blob/PREWHERE-row-filter/src/Interpreters/InterpreterSelectQuery.cpp#L410-L411
https://github.com/azat/ClickHouse/blob/PREWHERE-row-filter/src/Interpreters/InterpreterSelectQuery.cpp#L802-L803

**Comment by azat** [comment]
>Not working :\

What is the engine of the table?

**Comment by azat** [comment]
And can you provide an example?

**Comment by azat** [comment]
>https://github.com/azat/ClickHouse/blob/PREWHERE-row-filter/src/Interpreters/InterpreterSelectQuery.cpp#L410-L411
>https://github.com/azat/ClickHouse/blob/PREWHERE-row-filter/src/Interpreters/InterpreterSelectQuery.cpp#L802-L803

@filimonov This is ok (since `row_policy_filter` reseted before). The problem is in the backport.

File: tests/integration/test_row_policy_allow_insecure_prewhere/test.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+@pytest.fixture(scope="module", autouse=True)
+def started_cluster():
+    try:
+        cluster.start()
+
+        node.query('''
+            CREATE DATABASE mydb ENGINE=Ordinary;
+
+            CREATE TABLE mydb.prewhere_filter (a UInt8, b UInt8, c UInt8, s String) ENGINE MergeTree ORDER BY a PARTITION BY a SETTINGS index_granularity=1;
+            INSERT INTO mydb.prewhere_filter SELECT number, number, number, randomString(1000) FROM numbers(10);
```
Comment (filimonov) [inline tests/integration/test_row_policy_allow_insecure_prewhere/test.py:18]: For testing on 20.3: `randomString` => `randomPrintableASCII`

**Comment by filimonov** [comment]
it seems optimize_move_to_prewhere is still disabled: https://github.com/ClickHouse/ClickHouse/blob/cd372de4175ff38dbb52df54b363041aa3001275/src/Interpreters/InterpreterSelectQuery.cpp#L331

**Comment by azat** [comment]
>it seems optimize_move_to_prewhere is still disabled:

No, the `row_policy_filter` is cleaned earlier.

>Test

It, works, but a little bit dumb:

```sql
ANALYZE
SELECT *
FROM test
WHERE a = 1

┌─explain───────────────────┐
│ SELECT                    │
│     a,                    │
│     number                │
│ FROM test                 │
│ PREWHERE a = 1            │
│ WHERE (a = 1) AND (a = 1) │
└───────────────────────────┘
```

**Comment by azat** [comment]
>It, works, but a little bit dumb:

Not a bug, feature:
https://github.com/ClickHouse/ClickHouse/blob/18ef0c0c964530866f85fec3adb8ee8ac23b0b10/src/Interpreters/InterpreterSelectQuery.cpp#L426

P.S. backported to 20.3 also works

**Comment by filimonov** [comment]
Here is the test:
https://gist.github.com/filimonov/c84f48f91c155ab6afb91af33f49b084

**Comment by azat** [comment]
- the row-level security filter is in WHERE also because of https://github.com/ClickHouse/ClickHouse/blob/18ef0c0c964530866f85fec3adb8ee8ac23b0b10/src/Interpreters/InterpreterSelectQuery.cpp#L426
- expression from `WHERE` hasn't been pushed to `PREWHERE` because of `optimize_move_to_prewhere` does not works if `PREWHERE` already specified - fixed in 0acba8e5c697d5090f1061775160fdd659a3310b (2733dfe2a1540b1b4346452219b9766f2a0d6258)

**Comment by azat** [comment]
>expression from WHERE hasn't been pushed to PREWHERE because of optimize_move_to_prewhere does not works if PREWHERE already specified - fixed in 0acba8e (2733dfe)

This is not enough, take a look at f47a911935 / 2db3d6735b

**Comment by filimonov** [comment]
Now it fails with distributed :\ 

I've updated the test case: https://gist.github.com/filimonov/c84f48f91c155ab6afb91af33f49b084

**Comment by alex-zaitsev** [comment]
Since short circuit evaluation https://github.com/ClickHouse/ClickHouse/issues/12587 is not available, let's finalize this one since it is a critical performance feature for some users with row-level security.

**Comment by UnamedRus** [comment]
Double Distributed JOINs doesn't work correctly with row policy and allow_insecure_prewhere=1:

```
create table test_join
(
_time DateTime,
team LowCardinality(String) ,
dt  Date  DEFAULT toDate(_time),
hash         UInt64       DEFAULT cityHash64(concat(CAST(_time, 'String'), team)) 
)  
ENGINE = MergeTree PARTITION BY toDate(_time) ORDER BY (_time) SETTINGS index_granularity = 8192;


CREATE TABLE test_join_dist as test_join  ENGINE = Distributed(test_cluster_two_shards, default, test_join, hash);



insert into test_join(_time,team) values(now(),'EM');
insert into test_join(_time,team) values(now(),'EM');


SELECT *
FROM test_join_dist AS td
GLOBAL LEFT JOIN
(
    SELECT *
    FROM test_join_dist
    WHERE _time > '2020/10/12 00:00:00'
) AS td2 USING (team)
WHERE td._time > '2020/10/12 00:00:00'

┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-16 15:31:25 │ EM   │ 2020-10-16 │ 16042107814478222074 │ 0000-00-00 00:00:00 │ 0000-00-00 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘
┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-16 15:31:25 │ EM   │ 2020-10-16 │ 16042107814478222074 │ 0000-00-00 00:00:00 │ 0000-00-00 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘
┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-16 15:31:25 │ EM   │ 2020-10-16 │ 16042107814478222074 │ 0000-00-00 00:00:00 │ 0000-00-00 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘
┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-16 15:31:25 │ EM   │ 2020-10-16 │ 16042107814478222074 │ 0000-00-00 00:00:00 │ 0000-00-00 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘

4 rows in set. Elapsed: 0.014 sec.

cat /etc/clickhouse-server/users.d/b_user.xml
<yandex>
  <users>
    <b_user>
      <password></password>
      <networks incl="networks" replace="replace">
        <ip>::/0</ip>
      </networks>
      <profile>default</profile>
      <quota>default</quota>
      <allow_databases>
        <database>default</database>
      </allow_databases>
      <databases>
        <default>
          <test_join>
            <filter>team = 'EM'</filter>
          </test_join>
          <test_join_dist>
            <filter>team = 'EM'</filter>
          </test_join_dist>
        </default>
      </databases>
    </b_user>
  </users>
</yandex>
```

**Comment by azat** [comment]
>Double Distributed JOINs doesn't work correctly with row policy and allow_insecure_prewhere=1:

This is not strictly related to this PR, this is due to PREWHERE was broken for queries with GLOBAL IN, fixed here #15933 

```sql
-- example w/ allow_insecure_prewhere=0
-- and w/o row-levle security

SELECT *
FROM test_join_dist AS td
GLOBAL LEFT JOIN 
(
    SELECT *
    FROM test_join
    WHERE _time > '2020/10/12 00:00:00'
) AS td2 USING (team)
PREWHERE team = 'EM'
WHERE td._time > '2020/10/12 00:00:00'

Query id: fd4b90a7-1b45-4419-90ca-25f9f6bccb24

┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘
┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘

4 rows in set. Elapsed: 0.013 sec. 

ci-w1.bl.semrush.net :) show row policies;

SHOW ROW POLICIES

Query id: 97e64311-a3e2-42c6-969f-b2ac7e3dc6f7

Ok.

0 rows in set. Elapsed: 0.002 sec. 
```

**Comment by azat** [comment]
>Double Distributed JOINs doesn't work correctly with row policy and allow_insecure_prewhere=1:

This is not strictly related to this PR, this is due to PREWHERE was broken for queries with GLOBAL IN, fixed here #15933 

```sql
-- example w/ allow_insecure_prewhere=0
-- and w/o row-levle security

SELECT *
FROM test_join_dist AS td
GLOBAL LEFT JOIN 
(
    SELECT *
    FROM test_join
    WHERE _time > '2020/10/12 00:00:00'
) AS td2 USING (team)
PREWHERE team = 'EM'
WHERE td._time > '2020/10/12 00:00:00'

┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘
┌───────────────_time─┬─team─┬─────────dt─┬─────────────────hash─┬───────────td2._time─┬─────td2.dt─┬─td2.hash─┐
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
│ 2020-10-20 23:51:51 │ EM   │ 2020-10-20 │ 10344887503326349234 │ 1970-01-01 03:00:00 │ 1970-01-01 │        0 │
└─────────────────────┴──────┴────────────┴──────────────────────┴─────────────────────┴────────────┴──────────┘

4 rows in set. Elapsed: 0.013 sec. 

SHOW ROW POLICIES;

Ok.

0 rows in set. Elapsed: 0.002 sec. 
```

**Comment by alexey-milovidov** [comment]
We discussed internally the proper way of implementation for PREWHERE with Row Level Security, so this PR can be closed.

**Pull request closed without merging**